src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,591 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.text;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.Serial;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+import sun.util.locale.provider.LocaleProviderAdapter;
+
+/**
+ * {@code ListFormat} produces or parses a list of concatenated
+ * strings in a locale-sensitive way. Use {@code ListFormat} to construct a list
+ * of strings displayed for end users. For example, displaying 3 weekdays
+ * as "Monday, Wednesday, and Friday". This class provides the functionality
+ * defined in Unicode Consortium's LDML specification for
+ * <a href="https://www.unicode.org/reports/tr35/tr35-general.html#ListPatterns">
+ * List Patterns</a>.
+ * <p>
+ * Three types of concatenation are provided: {@link Type#STANDARD STANDARD},
+ * {@link Type#OR OR}, and {@link Type#UNIT UNIT}, also three styles for each
+ * type are provided: {@link Style#FULL FULL}, {@link Style#SHORT SHORT}, and
+ * {@link Style#NARROW NARROW}. The following snippet is an example of formatting
+ * the list of Strings {@code "Foo", "Bar", "Baz"} in US English with
+ * {@code STANDARD} type and {@code FULL} style:
+ * {@snippet lang=java :
+ * ListFormat.getInstance(Locale.US, ListFormat.Type.STANDARD, ListFormat.Style.FULL)
+ *     .format(List.of("Foo", "Bar", "Baz"))
+ * }
+ * This will produce the concatenated list string, "Foo, Bar, and Baz" as seen in
```

Comment:
With this sample code, if the Style is changed to SHORT, it would produce the same string. Would it be better to use the weekdays instead of Foo, Bar and Baz (as in the Unicode spec)? Esp. with the UNIT type, those examples explained it better, e.g. NARROW produces 3′ 7″.

Also, if the instance is of STANDARD/SHORT, does it format List.of("January", "February", "March") and return "Jan., Feb., and Mar.", or 3 feet, 7 inches to 3 ft, 7 in? The format method states simply "Returns the string that consists of the input strings, concatenated with the patterns of this ListFormat."  I wonder if it'd be helpful to explain a bit more or add one more sample.

Reply:
In fact, the sample in LDML's page seems to be incorrect. `standard-short` in English is defined as:
```
		<listPattern type="standard-short">
			<listPatternPart type="start">{0}, {1}</listPatternPart>
			<listPatternPart type="middle">{0}, {1}</listPatternPart>
			<listPatternPart type="end">{0}, &amp; {1}</listPatternPart>
			<listPatternPart type="2">{0} &amp; {1}</listPatternPart>
		</listPattern>
```
in `en.xml` file. So `&` is expected rather than `and`.

> Also, if the instance is of STANDARD/SHORT, does it format List.of("January", "February", "March") and return "Jan., Feb., and Mar.", or 3 feet, 7 inches to 3 ft, 7 in?

No it does not. The `format()` method does not alter the passed input strings, so it would not convert "January" to "Jan." even if `SHORT` Style is specified. I have added some extra explanations, that those patterns vary depending on the locale providers.

Reply:
Ah, I see. Those samples did give an impression that the format does more than concatenation.